src/dshould/ShouldType.d

```diff
@@ -0,0 +1,247 @@
+module dshould.ShouldType;
+
+import std.algorithm : map;
+import std.format : format;
+import std.range : iota;
+import std.string : empty, join;
+import std.traits : TemplateArgsOf;
+public import std.traits : isInstanceOf;
+import std.typecons : Tuple;
+
+public auto should(T)(lazy T got) pure
+{
+    T get() pure @safe
+    {
+        return got;
+    }
+
+    return ShouldType!(typeof(&get))(&get);
+}
+
+public struct ShouldType(G, string[] phrase = [])
+{
+    import std.algorithm : canFind;
+
+    public G got;
+
+    private int* refCount_ = null;
+
+    public auto addWord(string word)()
+    {
+        return ShouldType!(G, phrase ~ word)(this.got, this.refCount);
+    }
+
+    private this(G got) { this.got = got; this.refCount = 1; }
+
+    public this(G got, ref int refCount) @trusted
+    in
+    {
+        assert(refCount != CHAIN_TERMINATED, "don't copy Should that's been terminated");
+    }
+    do
+    {
+        this.got = got;
+        this.refCount_ = &refCount;
+        this.refCount++;
+    }
+
+    this(this) @trusted
+    in
+    {
+        assert(this.refCount != CHAIN_TERMINATED);
+    }
+    do
+    {
+        this.refCount++;
+    }
+
+    ~this() @trusted
+    {
+        this.refCount--;
+        assert(this.refCount > 0, "unterminated should-chain!");
+    }
+
+    public enum genCheck(string TestString, string[] ArgNames) = genCheckImpl(TestString, ArgNames);
+
+    private static string genCheckImpl(string testString, string[] argNames)
+    {
+        auto args = format!"%-(%s, %)"(argNames);
+        auto argStrings = format!"%(%s, %)"(argNames);
+
+        return format!q{{
+            import std.format : format;
+
+            check(mixin(format!q{%s}(%s)), format(": expected %s", %s), null, file, line);
+        }}(testString, argStrings, testString, args);
+    }
+
+    public void check(bool condition, lazy string left, lazy string right, string file, size_t line) pure @safe
+    {
+        terminateChain;
+
+        if (!condition)
+        {
+            throw new FluentException("test failed" ~ left, right, file, line);
+        }
+    }
+
+    public void allowOnlyWordsBefore(string[] allowedWords, string newWord)()
+    {
+        static foreach (word; phrase)
+        {
+            static assert(allowedWords.canFind(word), `bad grammar: "` ~ word ~ ` ` ~ newWord ~ `"`);
+        }
+    }
+
+    public void terminateChain()
```

Comment:
move `terminateChain` away, to keep related functions `allowOnlyWordsBefore` and `requireWord` together

Reply:
fixed